src/linking.rs

```diff
@@ -0,0 +1,45 @@
+use std::{path::Path, process::Command};
+
+use tempfile::TempDir;
+
+const EXIT_CODE_FAILURE: i32 = 1;
+const LINKER: &str = "rust-lld";
+
+pub fn link_normally(args: &[String]) -> Result<(), i32> {
+    let mut c = Command::new(LINKER);
+    c.args(args);
+    log::trace!("{:?}", c);
+
+    success_or_exitstatus(c)
+}
+
+pub fn link_again(
```

Comment:
I tried to make it more clear. What do you think?

...

...

... _(waiting for you to take a look)_ 

I'd like to put the inline comments into the doc-comment like following, but then we would have them duplicated, because I also like them to stay in the code, so one knows what each part does.
```rust
/// Re-link with modified arguments, which is the whole point of `flip-link`
///
/// **modifications**
/// * add the current dir to the linker search path to include all unmodified scripts there
/// * override `_stack_start` to make the stack start below fake RAM
/// * set working directory to temporary directory containing our new linker script,
///   this makes sure that it takes precedence over the original one
pub fn link_modified(
```

Reply:
Hrmm. I agree that the inline comments should stay in the code. In general, the function docs should give the reader an overview over the what and the why, but not necessarily the how: the steps we're taking to achieve what the function is supposed to do is irrevelant to the outside reader. So when you're saying "modified arguments", my first question as a user would be: "modified to achieve what?"

After looking at the code a bit more closely, I'd propose something like:

```
/// Link using a custom linker script and stack starting point. (This is the whole point of `flip-link)
///
/// `args` are arguments passed to the linker invocation
/// `current_dir` is the directory from which the linker was invoked
/// `stack_start` is the new, custom starting point from which our stack grows downwards–
///               this should be right *below* the `.bss+.data` region that we've moved to the top, e.g.:
///```
///  +-------------+
///  | .bss+.data  |
///  +-------------+ <-- `stack_start`
///  |    stack    | 
///  |      |      |
///  |      v      |
///  | ~~~~~~~~~~~ |
///  |             |
///  +-------------+
///```
///
/// `custom_linker_script_location` is the directory in which the linker script to be used is located
///
pub fn link_modified(
    args: &[String],
    current_dir: &Path,
    stack_start: u64,
    custom_linker_script_location: &TempDir,
) -> Result<(), i32> {
```
This probably needs a coherence check by @japaric or @jonas-schievink though.

Reply:
Adopted this with slight adaptions :)